File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.io.*;
import java.util.*;

public class Main {
	static int n, m;
	static boolean[] visited;
	static int[] sequence;
	static StringBuilder sb;
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine()); 
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        
        visited = new boolean[n+1]; // 1~n
        sequence = new int[m]; // 선택된 숫자  수열
        sb = new StringBuilder();
        backtrack(0);
        System.out.println(sb.toString());
    }
    
    public static void backtrack(int depth) {
    	if(depth == m) {
    		for(int num : sequence) {
    			sb.append(num).append(" ");
    		}
    		sb.deleteCharAt(sb.length() - 1);
    		sb.append("\n");
    		return;
    	}
    	
    	for(int i = 1; i <= n; i++) {
    		if(!visited[i]) {
    			visited[i] = true;
    			sequence[depth] = i;
    			backtrack(depth+1);
    			visited[i] = false;
    		}
    	}
    }
    
    
}

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.kt
package org.example

fun main() {
    val br = System.`in`.bufferedReader()
    val sb = StringBuilder()
    val (N, M) = br.readLine().split(" ").map { it.toInt() }
    // 1~N까지 자연수 중 중복 없이 M개 구하기
    val sequence = IntArray(M) // M개 고를 배열
    val visited = BooleanArray(N + 1) // 중복 없이, 0은 안 씀

    fun backtrack(depth: Int) {
        if (depth == M) {
            sequence.forEach {
                sb.append(it).append(" ")
            }
            sb.deleteCharAt(sb.length - 1)
            sb.append("\n")
            return
        }

        for (i in 1..N) {
            // 1~N까지의 수
            if (!visited[i]) {
                // 방문하지 않은 수라면
                visited[i] = true
                sequence[depth] = i
                backtrack(depth + 1)
                visited[i] = false
            }
        }
    }
    backtrack(0) //0부터 시작
    print(sb)
    
    br.close()
}
